ScreenLogic cleanups

dieselrabbit opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Bump screenlogicpy to 0.2.0.
Move heating functions from water_heater to climate platform.
Address notes from original PR.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->
N/A

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR is related to issue: (previous PR) https://github.com/home-assistant/core/pull/47933 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/17073

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

dieselrabbit committed:
ScreenLogic cleanup.
Bump screenlogicpy to 0.2.0.
Move heating functions from water_heater to climate platform.
Address notes from original PR.

bdraco commented:
Its not considered a breaking change since its unreleased

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,214 @@
+        """Supported features of the heater."""
+        return SUPPORTED_FEATURES
+
+    async def async_set_temperature(self, temperature, **kwargs) -> None:
```
```suggestion
    async def async_set_temperature(self, **kwargs) -> None:
```

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/da84d693517c00c49dcfdc9e074907e4c711c6d7

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,214 @@
+        return SUPPORTED_FEATURES
+
+    async def async_set_temperature(self, temperature, **kwargs) -> None:
+        """Change the setpoint of the heater."""
```
the schema for the set temperature service is

```
SET_TEMPERATURE_SCHEMA = vol.All(
    cv.has_at_least_one_key(
        ATTR_TEMPERATURE, ATTR_TARGET_TEMP_HIGH, ATTR_TARGET_TEMP_LOW
    ),
    make_entity_service_schema(
        {
            vol.Exclusive(ATTR_TEMPERATURE, "temperature"): vol.Coerce(float),
            vol.Inclusive(ATTR_TARGET_TEMP_HIGH, "temperature"): vol.Coerce(float),
            vol.Inclusive(ATTR_TARGET_TEMP_LOW, "temperature"): vol.Coerce(float),
            vol.Optional(ATTR_HVAC_MODE): vol.In(HVAC_MODES),
        }
    ),
)
```

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,214 @@
+        """Run when entity is about to be added."""
+        await super().async_added_to_hass()
+        _LOGGER.debug("Startup last preset is %s", self._last_preset)
+        if self._last_preset is None:
```
```suggestion
        if self._last_preset is not None:
            return
```

Then outdent the block below

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

bdraco commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -33,9 +38,9 @@ def name(self):
-        if device_class in DEVICE_CLASSES:
-            return device_class
+        device_class = self.sensor.get("device_type")
+        if device_class in SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS:
+            return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS[device_class]
         return None
```
```suggestion
        return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS.get(device_class)
```

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

bdraco commented on homeassistant/components/screenlogic/sensor.py:
```diff
@@ -42,9 +48,9 @@ def unit_of_measurement(self):
-        if device_class in DEVICE_CLASSES:
-            return device_class
+        device_class = self.sensor.get("device_type")
+        if device_class in SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS:
+            return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS[device_class]
         return None
```
```suggestion
        return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS.get(device_class)
```

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

bdraco commented on homeassistant/components/screenlogic/sensor.py:
```diff
@@ -86,9 +87,9 @@ def unit_of_measurement(self):
-        if device_class in DEVICE_CLASSES:
-            return device_class
+        device_class = self.pump_sensor.get("device_type")
+        if device_class in SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS:
+            return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS[device_class]
         return None
```
```suggestion
        return SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS.get(device_class)
```

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

bdraco commented on homeassistant/components/screenlogic/binary_sensor.py:
```diff
@@ -3,13 +3,18 @@
 
 _LOGGER = logging.getLogger(__name__)
 
+SL_DEVICE_TYPE_TO_HA_DEVICE_CLASS = {"alarm": DEVICE_CLASS_PROBLEM}
```
It would be a bit easier to maintain long term if `"alarm"` was a constant imported from `screenlogicpy` which would allow changes to the value without breaking anything

bdraco replied:
Same with the other sensor platform

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,214 @@
+        super().__init__(coordinator, body)
+        self._configured_heat_modes = []
+        # Is solar listed as available equipment?
+        if self.coordinator.data["config"]["equipment_flags"] & 0x1:
```
It would be cleaner to make `0x1` a constant imported from `screenlogicpy`

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/2d488e2a1747d24006a99a3b5f374c69f73748c7

dieselrabbit committed:
Fix temperature attribute

dieselrabbit committed:
Addressing notes.
Bump screenlogicpy to 0.2.1.
Made device_types constants.
Made (known) equipment flags constants.
Used dict.get() in places where None is the default.
Return fast with good _last_preset.

MartinHjelmare commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,217 @@
+        return f"{self.gateway_name} {ent_name}"
+
+    @property
+    def state(self) -> str:
```
Don't implement the state property. That's handled by the base climate entity.

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/4bfc1329668c67e3f776b7146547e4b0e58855a3

MartinHjelmare commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,217 @@
+
+    async def async_set_temperature(self, **kwargs) -> None:
+        """Change the setpoint of the heater."""
+        temperature = kwargs[ATTR_TEMPERATURE]
```
Temperature is not a required parameter for the climate set temperature service.

dieselrabbit replied:
https://github.com/home-assistant/core/pull/48136/commits/f077730806061630259f1bb29d7e6049203f806d

MartinHjelmare commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,217 @@
+        ):
+            await self.coordinator.async_request_refresh()
+        else:
+            _LOGGER.warning(
```
We usually raise an exception with the message as exception argument in service handler target methods. Here we could use HomeAssistantError.

dieselrabbit replied:
Should I use f-strings or `%s` with exceptions? Docs seem to indicate f-strings, but since (I believe) the exception message will be logged, I just wanted to confirm.

dieselrabbit replied:
Nevermind. Pretty clearly f-strings from looking at the codebase.

dieselrabbit replied:
https://github.com/home-assistant/core/pull/48136/commits/f077730806061630259f1bb29d7e6049203f806d

MartinHjelmare commented on tests/components/screenlogic/test_config_flow.py:
```diff
@@ -101,9 +101,33 @@ async def test_flow_discover_error(hass):
+    ), patch(
+        "homeassistant.components.screenlogic.config_flow.login.gateway_connect",
+        return_value="00-C0-33-01-01-01",
+    ):
```
Make sure to patch the integration setup functions on all create entry results to avoid setting up the whole integration.

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/df0a10b963e7a75fe5cfbf2df1d25f68948de543

MartinHjelmare commented on tests/components/screenlogic/test_config_flow.py:
```diff
@@ -213,9 +263,12 @@ async def test_option_flow(hass):
+    entry = MockConfigEntry(domain=DOMAIN)
     entry.add_to_hass(hass)
 
+    await hass.config_entries.async_setup(entry.entry_id)
```
Patch the integration setup functions or at least the library to avoid I/O.

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/df0a10b963e7a75fe5cfbf2df1d25f68948de543

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,217 @@
+        """Name of the heater."""
+        ent_name = self.body["heat_status"]["name"]
+        return f"{self.gateway_name} {ent_name}"
+
+    @property
+    def state(self) -> str:
+        """State of the heater."""
+        return self.hvac_mode
```
```suggestion
```

bdraco commented on homeassistant/components/screenlogic/climate.py:
```diff
@@ -0,0 +1,217 @@
+
+    async def async_set_temperature(self, **kwargs) -> None:
+        """Change the setpoint of the heater."""
+        temperature = kwargs[ATTR_TEMPERATURE]
```
```suggestion
        temperature = kwargs.get(ATTR_TEMPERATURE)
```

Probably raise if `temperature` is `None` since nothing else can happen here

dieselrabbit replied:
`raise` or `return`? I see quite a few integrations just `return` if `temperature` is `None`.

bdraco replied:
If it's not going to do what the user expects we should raise since they will not get any feedback otherwise.

dieselrabbit replied:
Commit https://github.com/home-assistant/core/pull/48136/commits/f077730806061630259f1bb29d7e6049203f806d

dieselrabbit committed:
Update homeassistant/components/screenlogic/climate.py

Let base entity handle state property.

Co-authored-by: J. Nick Koston <nick@koston.org>

dieselrabbit committed:
Patch integration setup functions.

dieselrabbit committed:
Exception, ATTR_TEMPERATURE notes

bdraco commented:
Looks good! Martin has a better eye for detail though so I'll wait for him to weigh in.

dieselrabbit commented:
Sounds good.  I greatly appreciate everyone's patience with me.

MartinHjelmare reviewed (approved):
Nice!
